dispatch.go

```diff
@@ -124,30 +124,33 @@ func (d *dispatch) handleCall(calls map[seqNumber]*call, c *call) {
 	seqid := d.nextSeqid()
 	c.seqid = seqid
 	calls[c.seqid] = c
-	err := d.dispatchMessage(c.ctx, MethodCall, seqid, c.method, c.arg)
-	if err != nil {
-		setResult := c.Finish(err)
-		if !setResult {
-			panic("c.Finish unexpectedly returned false")
-		}
-		return
-	}
+	errCh := d.dispatchMessage(c.ctx, MethodCall, seqid, c.method, c.arg)
 	d.log.ClientCall(seqid, c.method, c.arg)
 	go func() {
-		select {
-		case <-c.ctx.Done():
-			setResult := c.Finish(newCanceledError(c.method, seqid))
-			if !setResult {
-				// The result of the call has already
-				// been processed.
-				return
+		for {
```

Comment:
you have a for loop, but an unconditional return at the end, so no need for the for loop.